feat: add vim types

superle3 opened the pull request:
It seems Obsidians version is based off  https://github.com/replit/codemirror-vim/commit/0ac05aba9ef84efb236d8c686f8699ef4295b200 with slight changes, not to the api but behaviour. 
But that version has barely any types
 and from a quick glance the api hasn't changed, but slight additions have been made and types have been added thus i am basing the types off their tag v6.3.0
 
Resolved #132.

mnaoumov committed:
chore(release): 3.9.5

Fevol committed:
docs: add new tools

Fevol committed:
docs: add new tool

mnaoumov committed:
chore: update libs

mnaoumov committed:
fix: patch @antfu/utils

mnaoumov committed:
chore(release): 3.9.6

mnaoumov committed:
feat: Menu.forEvent

mnaoumov committed:
chore(release): 3.10.0

ElsaTam committed:
Add documentation and fix some errors

mnaoumov committed:
feat: Add documentation for the graph and fix some errors #121

mnaoumov committed:
docs: rephrase

mnaoumov committed:
chore: format

mnaoumov committed:
chore(release): 3.11.0

mnaoumov committed:
feat: WorkspaceLeaf/WorkspaceItem

mnaoumov committed:
chore(release): 3.12.0

mnaoumov committed:
chore: fix version for patched dependency

mnaoumov committed:
chore: update libs

mnaoumov committed:
chore: update libs

mnaoumov committed:
chore(release): 3.12.1

mnaoumov committed:
fix: PropertyWidget

mnaoumov committed:
chore(release): 3.13.0

mnaoumov committed:
fix: PropertyWidget

mnaoumov committed:
chore(release): 3.13.1

mnaoumov committed:
fix: PropertyWidget. Support both versions:

public and catalyst

mnaoumov committed:
chore(release): 3.13.2

mnaoumov committed:
fix: Use precise versions of dep libs (#109)

mnaoumov committed:
chore(release): 3.13.3

mnaoumov committed:
fix: missing patch

mnaoumov committed:
chore(release): 3.13.4

mnaoumov committed:
fix: update codemirror packages

as it's unclear which ones to take

mnaoumov committed:
chore(release): 3.13.5

kotaindah55 committed:
feat: extend `isUserEvent()` with overloads

kotaindah55 committed:
docs: add missing description

mnaoumov committed:
feat: add property widget components

mnaoumov committed:
feat: extend isUserEvent() with overloads #122

mnaoumov committed:
chore(release): 3.14.0

mnaoumov committed:
fix: allow null to TextProperty

mnaoumov committed:
chore(release): 3.14.1

mnaoumov committed:
feat: AbstractInputSuggest / PopoverSuggest / SuggestionContainer

mnaoumov committed:
chore(release): 3.15.0

mnaoumov committed:
feat: EmberContext

mnaoumov committed:
docs: Embed

mnaoumov committed:
chore: format

mnaoumov committed:
chore(release): 3.16.0

mnaoumov committed:
fix: include menu separator

Discovered in https://github.com/mnaoumov/obsidian-root-folder-context-menu/pull/4

mnaoumov committed:
chore(release): 3.16.1

mnaoumov committed:
fix: watch typings

mnaoumov committed:
chore(release): 3.16.2

mnaoumov committed:
fix: build

mnaoumov committed:
chore(release): 3.16.3

mnaoumov committed:
fix: revert incorrect watch changes

mnaoumov committed:
chore(release): 3.16.4

mnaoumov committed:
docs: fix misleading comment

mnaoumov committed:
chore(release): 3.16.5

mnaoumov committed:
fix: name() type

mnaoumov committed:
chore(release): 3.16.6

mnaoumov committed:
docs: short README

mnaoumov committed:
docs: short README

mnaoumov committed:
docs: fix alt

mnaoumov committed:
docs: link to branch

mnaoumov committed:
docs: fix typo

mnaoumov committed:
docs: make link

mnaoumov committed:
fix: remove 1.9.0 code

mnaoumov committed:
fix: remove deprecated in 1.9.0

mnaoumov committed:
chore: cleanup deprecated

mnaoumov committed:
chore: cleanup deprecated

mnaoumov committed:
chore: rename stable -> public

mnaoumov committed:
docs: add public suffix

mnaoumov committed:
chore: Push Branch (Proxy)

mnaoumov committed:
chore: test empty

mnaoumov committed:
chore: clean .github

mnaoumov committed:
chore: rename proxy

mnaoumov committed:
chore: Ignore .github/scripts

mnaoumov committed:
chore: test empty

mnaoumov committed:
chore: test empty

mnaoumov committed:
chore: test empty

mnaoumov committed:
chore: test empty

mnaoumov committed:
chore: rename main-workflow

mnaoumov committed:
chore: test empty

mnaoumov committed:
chore: test empty

mnaoumov committed:
chore: test empty

mnaoumov committed:
chore: allow write

github-actions[bot] committed:
chore: generate README.md from template

mnaoumov committed:
chore: test empty

mnaoumov committed:
chore: test empty

github-actions[bot] committed:
chore: generate README.md from template

mnaoumov committed:
chore: test empty

mnaoumov committed:
chore: move to workflow-scripts

mnaoumov committed:
chore: test empty

mnaoumov committed:
chore: update urls

mnaoumov committed:
Merge branch 'release/obsidian-public/1.8.10' into release/obsidian-catalyst/1.9.9

mnaoumov committed:
chore: test empty

github-actions[bot] committed:
chore: test empty

github-actions[bot] committed:
chore: generate README.md from template

github-actions[bot] committed:
chore: generate README.md from template

github-actions[bot] committed:
docs: update changelog link

mnaoumov committed:
chore: add single release tag

mnaoumov committed:
chore: rename back to Proxy

mnaoumov committed:
chore: add proxy workflows

mnaoumov committed:
chore: use annotated tag

mnaoumov committed:
chore: release-beta

mnaoumov committed:
chore: force update tags

github-actions[bot] committed:
chore(release): 4.1.0

github-actions[bot] committed:
chore(release): 4.2.0

github-actions[bot] committed:
chore(release): 4.3.0

github-actions[bot] committed:
chore(release): 4.4.0

github-actions[bot] committed:
chore(release): 4.5.0

github-actions[bot] committed:
chore(release): 4.6.0

mnaoumov committed:
chore: fix build

github-actions[bot] committed:
chore(release): 4.7.0

github-actions[bot] committed:
chore(release): 4.8.0

mnaoumov committed:
chore: fix build

mnaoumov committed:
chore: sort ignores

mnaoumov committed:
chore: exclude workflow-scripts from npm

mnaoumov committed:
chore: fix warnings

mnaoumov committed:
chore: suppress misleading TS2411 warning

github-actions[bot] committed:
chore(release): 4.9.0

github-actions[bot] committed:
chore(release): 4.10.0

mnaoumov committed:
Merge branch 'release/obsidian-public/1.8.10' into release/obsidian-catalyst/1.9.9

github-actions[bot] committed:
chore(release): 4.11.0

github-actions[bot] committed:
chore(release): 4.12.0

mnaoumov committed:
Merge branch 'release/obsidian-catalyst/1.9.9' into release/obsidian-catalyst/1.9.10

github-actions[bot] committed:
chore(release): 4.13.0

github-actions[bot] committed:
chore: generate README.md from template

mnaoumov committed:
chore: fix url

github-actions[bot] committed:
chore(release): 4.14.0

github-actions[bot] committed:
chore(release): 4.15.0

mnaoumov committed:
chore: reuse script from main

github-actions[bot] committed:
chore(release): 4.16.0

github-actions[bot] committed:
chore(release): 4.17.0

github-actions[bot] committed:
chore(release): 4.18.0

mnaoumov committed:
chore: explicitly list all packaged files

github-actions[bot] committed:
chore(release): 4.19.0

mnaoumov committed:
Merge branch 'release/obsidian-public/1.8.10' into release/obsidian-catalyst/1.9.9

# Conflicts:
#	package-lock.json
#	package.json

mnaoumov committed:
Merge branch 'release/obsidian-catalyst/1.9.9' into release/obsidian-catalyst/1.9.10

# Conflicts:
#	package-lock.json
#	package.json

mnaoumov committed:
Merge branch 'release/obsidian-catalyst/1.9.10' into release/obsidian-public/1.9.10

github-actions[bot] committed:
chore(release): 4.21.0

github-actions[bot] committed:
chore(release): 4.22.0

mnaoumov committed:
fix: receive-files-menu / receive-text-menu

github-actions[bot] committed:
chore(release): 4.24.0

Commit:
Updated MetadataTypeManager

Commit:
Added 'property' type widget'

github-actions[bot] committed:
chore: generate README.md from template

mnaoumov committed:
docs: changelog url

Commit:
Updated PropertyInfo

Commit:
Removed metadataEditor property from PropertyRenderContext

Commit:
PropertyComponentBase does not extend Component

Commit:
Removed unused import

Commit:
Created PropertyPropertyWidgetComponent and updated entry in MetadataTypeManagerRegisteredTypeWidgetsRecord

github-actions[bot] committed:
chore(release): 4.25.0

mnaoumov committed:
feat: Various MetadataTypeManager related updates #128

mnaoumov committed:
chore: format

mnaoumov committed:
chore: fix .js imports

mnaoumov committed:
chore: format

mnaoumov committed:
feat: fuzzySearch/prepareQuery

mnaoumov committed:
chore/docs: reformat optional callback

mnaoumov committed:
feat: revisit all PropertyWidgetComponents

mnaoumov committed:
Merge branch 'release/obsidian-public/1.9.10' into release/obsidian-catalyst/1.9.11

# Conflicts:
#	README.md
#	package-lock.json
#	package.json

github-actions[bot] committed:
chore(release): 4.26.0

github-actions[bot] committed:
chore(release): 4.27.0

mnaoumov committed:
chore: fix double comments

mnaoumov committed:
chore: format

mnaoumov committed:
feat: merge editors

mnaoumov committed:
Merge branch 'release/obsidian-public/1.9.10' into release/obsidian-catalyst/1.9.11

# Conflicts:
#	package-lock.json
#	package.json

github-actions[bot] committed:
chore(release): 4.28.0

github-actions[bot] committed:
chore(release): 4.29.0

mnaoumov committed:
chore: spellcheck

mnaoumov committed:
fix: missing moment import

mnaoumov committed:
Merge branch 'release/obsidian-public/1.9.10' into release/obsidian-catalyst/1.9.11

github-actions[bot] committed:
chore(release): 4.30.0

github-actions[bot] committed:
chore(release): 4.31.0

Commit:
Updated

mnaoumov committed:
fix: render/data add null type

github-actions[bot] committed:
chore: generate README.md from template

mnaoumov committed:
docs: changelog url

github-actions[bot] committed:
chore: generate README.md from template

mnaoumov committed:
docs: update url

mnaoumov committed:
fix: PropertyWidget/render always accepts null

github-actions[bot] committed:
chore(release): 4.32.0

github-actions[bot] committed:
chore(release): 4.33.0

github-actions[bot] committed:
chore(release): 4.34.0

mnaoumov committed:
feat: Updated MetadataTypeManager.getTypeInfo() #130

mnaoumov committed:
feat: cleanup GetTypeInfoOptions

mnaoumov committed:
Merge branch 'release/obsidian-public/1.9.10' into release/obsidian-catalyst/1.9.11

# Conflicts:
#	package-lock.json
#	package.json

mnaoumov committed:
Merge branch 'release/obsidian-catalyst/1.9.11' into release/obsidian-catalyst/1.9.12

mnaoumov committed:
Merge branch 'release/obsidian-catalyst/1.9.12' into release/obsidian-public/1.9.12

# Conflicts:
#	package-lock.json
#	package.json

github-actions[bot] committed:
chore(release): 4.38.0

mnaoumov committed:
fix: don't assume type in widgets

github-actions[bot] committed:
chore(release): 4.39.0

mnaoumov committed:
chore: format

mnaoumov committed:
fix: more bases-related

github-actions[bot] committed:
chore(release): 4.40.0

mnaoumov committed:
feat: more bases

mnaoumov committed:
chore: cleanup

github-actions[bot] committed:
chore(release): 4.41.0

mnaoumov committed:
fix: data type

github-actions[bot] committed:
chore(release): 4.42.0

mnaoumov committed:
feat: more bases

github-actions[bot] committed:
chore(release): 4.43.0

mnaoumov committed:
fix: wrong data type

github-actions[bot] committed:
chore(release): 4.44.0

mnaoumov committed:
feat: GetRecentFilesOptions

mnaoumov committed:
feat: BookmarkItem

mnaoumov committed:
feat: SuggestModalChooser

github-actions[bot] committed:
chore(release): 4.45.0

mnaoumov committed:
fix: GetRecentFilesOptions

github-actions[bot] committed:
chore(release): 4.46.0

mnaoumov committed:
fix: BookmarkItem

github-actions[bot] committed:
chore(release): 4.47.0

mnaoumov committed:
feat: ImportedAttachment

mnaoumov committed:
feat: ImportedAttachment

github-actions[bot] committed:
chore(release): 4.48.0

Commit:
Updated MultiSelect

mnaoumov committed:
feat: Updated MultiSelect #131

mnaoumov committed:
docs: add format

github-actions[bot] committed:
chore(release): 4.49.0

mnaoumov committed:
fix: patch @types/css-font-loading-module

github-actions[bot] committed:
chore(release): 4.50.0

mnaoumov committed:
feat: Command

mnaoumov committed:
chore: format

github-actions[bot] committed:
chore: generate README.md from template

mnaoumov committed:
chore: fix url

github-actions[bot] committed:
chore: generate README.md from template

mnaoumov committed:
chore: url

mnaoumov committed:
chore: merge public 1.9.12

github-actions[bot] committed:
chore(release): 4.53.0

mnaoumov committed:
chore: enable conventional commits

mnaoumov committed:
chore: update libs

github-actions[bot] committed:
chore: generate README.md from template

mnaoumov committed:
docs: update changelog

mnaoumov committed:
fix: build

mnaoumov committed:
fix: build

github-actions[bot] committed:
chore(release): 4.54.0

mnaoumov committed:
fix: add missing API and @since

mnaoumov committed:
chore: format

mnaoumov committed:
docs: parseFromString

mnaoumov committed:
fix: add missing import

mnaoumov committed:
docs: sort tags

mnaoumov committed:
fix: build

mnaoumov committed:
feat: include since config

github-actions[bot] committed:
chore(release): 4.56.0

mnaoumov committed:
chore: spellcheck

github-actions[bot] committed:
chore(release): 4.58.0

mnaoumov committed:
chore: update obsidian dependency

github-actions[bot] committed:
chore(release): 4.61.0

superle3 committed:
feat: added compiled vim types

added the result of  `tsc --declaration --emitDeclarationOnly --outFile types.d.ts -p ./tsconfig.json`
from replit/codemirror of tag v6.3.0.
Obsidian's version seems to be mostly adapted from https://github.com/replit/codemirror-vim/commit/0ac05aba9ef84efb236d8c686f8699ef4295b200 but that version doesn't have types and the changes seem to be mostly bug fixes small features and small rewrites but no api changes.

superle3 committed:
feat(vim types): added vim types based on the compiled types of replit/codemirror-vim tag v6.3.0

/** @todo Documentation incomplete. */

fix #123

superle3 committed:
feat(vim types): added vim types based on the compiled types of replit/codemirror-vim tag v6.3.0

Adapted the compiled types from replit/codemirror-vim into the existing Obsidian typings structure.

fix [#123](https://github.com/superle3/obsidian-typings/issues/123)

superle3 commented:
How do you want me to handle the documentation of https://codemirror.net/5/doc/manual.html#vimapi., just copy it as is or do some processing.

Same goes for the compiled `types.d.ts` is it good as it is now or does it need to be different.

(will squash the commits later to one commit following the convention, but for now keeping them expanded)

mnaoumov commented on src/@codemirror__view/internals/Vim/VimApi.d.ts:
```diff
@@ -1,5 +1,7 @@
+import type { ActionArgsPartial } from 'types';
+// import type { VimEditor } from './VimEditor.d.ts';
+// import type { VimState } from './VimState.d.ts';
+import type { VimStateVim as vimState } from './VimStateVim.js';
```
use aliases only when they are conflicting with the typings you are trying to write

mnaoumov commented on src/@codemirror__view/internals/Vim/VimApi.d.ts:
```diff
@@ -8,92 +10,475 @@ import type { VimState } from './VimState.d.ts';
     /** @todo Documentation incomplete. */
+    /**
+     * Whether to use suppress the use of console.log when catching an error in the function returned by `this.findKey`.
+     * Defaults to false.
```
Use 

```
`false`
```

mnaoumov commented on src/@codemirror__view/internals/Vim/VimApi.d.ts:
```diff
@@ -8,92 +10,475 @@ import type { VimState } from './VimState.d.ts';
 
     /** @todo Documentation incomplete. */
-    buildKeyMap(): void;
+    defineAction: (name: string, fn: import('types').ActionFn) => void;
```
Use function syntax when possible

```ts
defineAction(name: string, fn: ActionFn): void;
```

mnaoumov commented on src/@codemirror__view/internals/Vim/VimApi.d.ts:
```diff
@@ -8,92 +10,475 @@ import type { VimState } from './VimState.d.ts';
 
     /** @todo Documentation incomplete. */
-    defineOperator(name: unknown, fn: unknown): unknown;
+    defineOption: {
```
here should be three overloads

mnaoumov commented on src/@codemirror__view/internals/Vim/VimApi.d.ts:
```diff
@@ -8,92 +10,475 @@ import type { VimState } from './VimState.d.ts';
     /** @todo Documentation incomplete. */
-    handleKey(cm: unknown, key: unknown, origin: unknown): unknown;
+    InsertModeKey: {
+        new (keyName: string, e: KeyboardEvent): {
```
return type should be extracted as a separate interface

mnaoumov commented on src/@codemirror__view/internals/Vim/types.d.ts:
```diff
@@ -0,0 +1,1719 @@
+declare module 'types' {
```
No meaningless modules. We add modules only augment existing public libraries

mnaoumov commented:
> > One entity (class/interface/type/enum/...) per file.
> 
> how do you want me to do that? create a `types` folder inside the `Vim` folder and somehow use the typescript lsp to refactor all types to their own file?

I don't see what value `types` folder brings. But, yes, every type in its own file. You can group them in folders if it provides logical sense.

> > Add links with proofs for everything like
> > It seems Obsidians version is based off [replit/codemirror-vim@0ac05ab](https://github.com/replit/codemirror-vim/commit/0ac05aba9ef84efb236d8c686f8699ef4295b200)
> 
> in each file? and just the commit or do you expect me to hunt down each method/type definition with a permalink.

No, of course not in all files. Just the logical top one, the root of all `Vim` typings.

> 
> > No anonymous entities.
> 
> What does this mean? you mean to long inline types? like
> 
> ```ts
>     getRegisterController: () => {
>         registers: {
>             [x: string]: {
>                 keyBuffer: string[];
>                 insertModeChanges: import('types').InsertModeChanges[];
>                 searchQueries: string[];
>                 linewise: boolean;
>                 blockwise: boolean;
>                 setText(
>                     text?: string | undefined,
>                     linewise?: boolean | undefined,
>                     blockwise?: boolean | undefined
>                 ): void;
>                 pushText(text: string, linewise?: boolean | undefined): void;
>                 pushInsertModeChanges(
>                     changes: import('types').InsertModeChanges
>                 ): void;
>                 pushSearchQuery(query: string): void;
>                 clear(): void;
>                 toString(): string;
>             };
>         };
> // further overloads ...
> }
> ```

Yeah, like that. In this project we name every single anonymous interface type for argument types and for the return types.

```ts
key: {
  prop: string;
};

// becomes

key: Key;

interface Key {
  prop: string;
}
```

Sometimes we even extract union type if it makes sense

```ts
key: A | B | C;

// becomes

key: Key;

type Key = A | B | C;
```

> 
> > Docs https://codemirror.net/5/doc/manual.html#vimapi has to be linked accordingly (see point 3)
> > But it also has to be converted into proper tsdocs.
> 
> I should have been more specific, the structure of as is I mean is
> 
> ```
> /*
>  * <insert-copied-explanation-from-docs>
>  * @arg ... <explanation>
>  * ...
>  * @returns <explanation
> */
> ```
> 
> As there a lot of types and I am not that creative/good at writing.

You don't have to be. Just add

```ts
@todo Documentation incomplete.
```

to each member you are not willing to write up.

> 
> > I will provide random reviews (not full) to pinpoint the issues.
> 
> I know the problems `types.d.ts` is fully undocumented and has to split apart and `VimStateVim.d.ts` needs it tsdoc back. I needed an answer before wasting a lot of time formatting it in a way that wasn't desirable, i'll put it back to draft.

Feel free to ask me if the guidelines are not clear. We are not trying to be assholes, we just try to keep the typings on a certain level so people benefit maximum from it.

Also learn about

```
npm run format
```

It will take a while. Don't judge the intermediate result until the script is done.

I also recommend to commit your changes BEFORE running the formatting and review its results AFTER, so you can be sure it doesn't mess up.

superle3 commented:
> I don't see what value types folder brings. But, yes, every type in its own file. You can group them in folders if it provides logical sense.

The `types.d.ts` file is the compiled types version that you get when running `tsc --declaration ....` in the cm-vim repo, then I copied the types out of them into their corresponding files. folder could also be "viminternal" or smth, but its to indicate that the root files are the main api.

superle3 committed:
refactor(extraction types): extracted each module and type into its own folder/file

following style guide specifying each type deserves its own file.
This still includes inline types though, in the next commit.

superle3 committed:
refactor(extraction types): extracting inline types into their own types

mnaoumov commented:
@superle3 is it still draft?
